File: packages/contracts/src/catalog-bridge.ts
import { z } from "zod";

export const CATALOG_BRIDGE_PROTOCOL_VERSION = "2026-07-bridge-v1" as const;
export const CATALOG_BRIDGE_MAX_BATCH_RECORDS = 250;

const ForbiddenFieldTokens = new Set([
  "customer",
  "customers",
  "order",
  "orders",
  "payment",
  "payments",
  "password",
  "passwords",
  "session",
  "sessions",
  "token",
  "tokens",
  "address",
  "addresses",
]);

function containsForbiddenFieldToken(value: string): boolean {
  const normalized = value.replace(/([a-z0-9])([A-Z])/g, "$1_$2").toLowerCase();
  return normalized
    .split(/[_\-.]+/)
    .some((token) => ForbiddenFieldTokens.has(token));
}

const SafeFieldKeySchema = z
  .string()
  .min(1)
  .max(160)
  .refine((value) => !containsForbiddenFieldToken(value), {
    message: "Sensitive or non-catalog field names are not allowed",
  });

export type CatalogBridgeJsonValue =
  | string
  | number
  | boolean
  | null
  | CatalogBridgeJsonValue[]
  | { [key: string]: CatalogBridgeJsonValue };

export const CatalogBridgeJsonValueSchema: z.ZodType<CatalogBridgeJsonValue> =
  z.lazy(() =>
    z.union([
      z.string().max(20_000),
      z.number().finite(),
      z.boolean(),
      z.null(),
      z.array(CatalogBridgeJsonValueSchema).max(500),
      z.record(SafeFieldKeySchema, CatalogBridgeJsonValueSchema),
    ]),
  );

export const CatalogBridgeSeoSchema = z
  .object({
    title: z.string().max(500).nullable().optional(),
    description: z.string().max(2_000).nullable().optional(),
  })
  .strict();

export const CatalogBridgeImageSchema = z
  .object({
    url: z.string().url().max(2_000),
    altText: z.string().max(1_000).nullable().optional(),
    position: z.number().int().nonnegative().max(10_000).optional(),
    variantId: z.string().min(1).max(500).nullable().optional(),
  })
  .strict();

export const CatalogBridgeVariantSchema = z
  .object({
    id: z.string().min(1).max(500),
    sku: z.string().max(500).nullable().optional(),
    title: z.string().max(1_000).nullable().optional(),
    price: z.string().max(100).nullable().optional(),
    compareAtPrice: z.string().max(100).nullable().optional(),
    currency: z.string().length(3).nullable().optional(),
    availability: z.string().max(100).nullable().optional(),
    options: z
      .record(SafeFieldKeySchema, CatalogBridgeJsonValueSchema)
      .optional(),
    attributes: z
      .record(SafeFieldKeySchema, CatalogBridgeJsonValueSchema)
      .optional(),
    images: z.array(CatalogBridgeImageSchema).max(100).optional(),
    createdAt: z.string().datetime().nullable().optional(),
    updatedAt: z.string().datetime().nullable().optional(),
  })
  .strict();

export const CatalogBridgeProductSchema = z
  .object({
    id: z.string().min(1).max(500),
    title: z.string().min(1).max(1_000),
    description: z.string().max(100_000).nullable().optional(),
    slug: z.string().max(1_000).nullable().optional(),
    url: z.string().url().max(2_000).nullable().optional(),
    canonicalUrl: z.string().url().max(2_000).nullable().optional(),
    status: z.string().max(100).nullable().optional(),
    brand: z.string().max(500).nullable().optional(),
    productType: z.string().max(500).nullable().optional(),
    categories: z.array(z.string().min(1).max(500)).max(100).optional(),
    collections: z.array(z.string().min(1).max(500)).max(100).optional(),
    tags: z.array(z.string().min(1).max(500)).max(500).optional(),
    attributes: z
      .record(SafeFieldKeySchema, CatalogBridgeJsonValueSchema)
      .optional(),
    seo: CatalogBridgeSeoSchema.optional(),
    images: z.array(CatalogBridgeImageSchema).max(500).optional(),
    variants: z.array(CatalogBridgeVariantSchema).max(2_000).optional(),
    createdAt: z.string().datetime().nullable().optional(),
    updatedAt: z.string().datetime().nullable().optional(),
  })
  .strict()
  .superRefine((product, context) => {
    const variantIds = new Set<string>();
    for (const [index, variant] of (product.variants ?? []).entries()) {
      if (variantIds.has(variant.id)) {
        context.addIssue({
          code: "custom",
          path: ["variants", index, "id"],
          message: "Variant IDs must be unique inside a product",
        });
      }
      variantIds.add(variant.id);
    }
    for (const [index, image] of (product.images ?? []).entries()) {
      if (image.variantId && !variantIds.has(image.variantId)) {
        context.addIssue({
          code: "custom",
          path: ["images", index, "variantId"],
          message:
            "Image variantId must reference a variant in the same product",
        });
      }
    }
  });

export const CatalogBridgeSnapshotManifestSchema = z
  .object({
    protocolVersion: z.literal(CATALOG_BRIDGE_PROTOCOL_VERSION),
    snapshotId: z.string().uuid(),
    startedAt: z.string().datetime(),
    declaredProductCount: z.number().int().nonnegative(),
    declaredVariantCount: z.number().int().nonnegative(),
    sourceLabel: z.string().min(1).max(200).optional(),
    metadata: z.record(SafeFieldKeySchema, z.string().max(500)).optional(),
  })
  .strict();

export const CatalogBridgeBatchSchema = z
  .object({
    protocolVersion: z.literal(CATALOG_BRIDGE_PROTOCOL_VERSION),
    snapshotId: z.string().uuid(),
    sequence: z.number().int().nonnegative(),
    records: z
      .array(CatalogBridgeProductSchema)
      .min(1)
      .max(CATALOG_BRIDGE_MAX_BATCH_RECORDS),
  })
  .strict();

export const CatalogBridgeCompleteRequestSchema = z
  .object({
    protocolVersion: z.literal(CATALOG_BRIDGE_PROTOCOL_VERSION),
    snapshotId: z.string().uuid(),
    batchCount: z.number().int().positive(),
    productCount: z.number().int().nonnegative(),
    variantCount: z.number().int().nonnegative(),
  })
  .strict();

export const CatalogBridgeSourceProvisionResponseSchema = z
  .object({
    sourceId: z.string().uuid(),
    endpoint: z.string().url(),
    token: z.string().min(32),
    tokenFingerprint: z.string().length(12),
    protocolVersion: z.literal(CATALOG_BRIDGE_PROTOCOL_VERSION),
  })
  .strict();

export const CatalogBridgeSnapshotStatusSchema = z
  .object({
    sourceId: z.string().uuid(),
    snapshotId: z.string().uuid(),
    status: z.enum([
      "receiving",
      "queued",
      "processing",
      "completed",
      "failed",
    ]),
    acceptedBatches: z.number().int().nonnegative(),
    acceptedProducts: z.number().int().nonnegative(),
    acceptedVariants: z.number().int().nonnegative(),
    ingestionJobId: z.string().uuid().nullable(),
  })
  .strict();

export type CatalogBridgeProduct = z.infer<typeof CatalogBridgeProductSchema>;
export type CatalogBridgeVariant = z.infer<typeof CatalogBridgeVariantSchema>;
export type CatalogBridgeImage = z.infer<typeof CatalogBridgeImageSchema>;
export type CatalogBridgeSnapshotManifest = z.infer<
  typeof CatalogBridgeSnapshotManifestSchema
>;
export type CatalogBridgeBatch = z.infer<typeof CatalogBridgeBatchSchema>;
export type CatalogBridgeCompleteRequest = z.infer<
  typeof CatalogBridgeCompleteRequestSchema
>;
export type CatalogBridgeSourceProvisionResponse = z.infer<
  typeof CatalogBridgeSourceProvisionResponseSchema
>;
export type CatalogBridgeSnapshotStatus = z.infer<
  typeof CatalogBridgeSnapshotStatusSchema
>;

File: apps/api/catora_api/schemas/catalog_bridge.py
from __future__ import annotations

import re
import uuid
from datetime import datetime
from typing import Any, Self

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator, model_validator

CATALOG_BRIDGE_PROTOCOL_VERSION = "2026-07-bridge-v1"
CATALOG_BRIDGE_MAX_BATCH_RECORDS = 250
_FORBIDDEN_FIELD = re.compile(
    r"(^|[_\-.])(customer|customers|order|orders|payment|payments|password|passwords|"
    r"session|sessions|token|tokens|address|addresses)([_\-.]|$)",
    re.IGNORECASE,
)


class CatalogBridgeModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


JsonValue = dict[str, Any] | list[Any] | str | int | float | bool | None


def _validate_catalog_value(value: JsonValue, *, depth: int = 0) -> JsonValue:
    if depth > 8:
        raise ValueError("Catalog attribute nesting cannot exceed 8 levels")
    if isinstance(value, dict):
        if len(value) > 500:
            raise ValueError("Catalog attribute objects cannot exceed 500 fields")
        for key, item in value.items():
            if not key or len(key) > 160 or _FORBIDDEN_FIELD.search(key):
                raise ValueError(f"Catalog attribute key '{key}' is not allowed")
            _validate_catalog_value(item, depth=depth + 1)
    elif isinstance(value, list):
        if len(value) > 500:
            raise ValueError("Catalog attribute arrays cannot exceed 500 items")
        for item in value:
            _validate_catalog_value(item, depth=depth + 1)
    elif isinstance(value, str) and len(value) > 20_000:
        raise ValueError("Catalog attribute strings cannot exceed 20000 characters")
    return value


class CatalogBridgeSeo(CatalogBridgeModel):
    title: str | None = Field(default=None, max_length=500)
    description: str | None = Field(default=None, max_length=2_000)


class CatalogBridgeImage(CatalogBridgeModel):
    url: HttpUrl
    alt_text: str | None = Field(default=None, alias="altText", max_length=1_000)
    position: int | None = Field(default=None, ge=0, le=10_000)
    variant_id: str | None = Field(default=None, alias="variantId", max_length=500)


class CatalogBridgeVariant(CatalogBridgeModel):
    id: str = Field(min_length=1, max_length=500)
    sku: str | None = Field(default=None, max_length=500)
    title: str | None = Field(default=None, max_length=1_000)
    price: str | None = Field(default=None, max_length=100)
    compare_at_price: str | None = Field(default=None, alias="compareAtPrice", max_length=100)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    availability: str | None = Field(default=None, max_length=100)
    options: dict[str, JsonValue] = Field(default_factory=dict, max_length=200)
    attributes: dict[str, JsonValue] = Field(default_factory=dict, max_length=500)
    images: list[CatalogBridgeImage] = Field(default_factory=list, max_length=100)
    created_at: datetime | None = Field(default=None, alias="createdAt")
    updated_at: datetime | None = Field(default=None, alias="updatedAt")

    @field_validator("options", "attributes")
    @classmethod
    def validate_attributes(cls, value: dict[str, JsonValue]) -> dict[str, JsonValue]:
        return _validate_catalog_value(value)  # type: ignore[return-value]


class CatalogBridgeProduct(CatalogBridgeModel):
    id: str = Field(min_length=1, max_length=500)
    title: str = Field(min_length=1, max_length=1_000)
    description: str | None = Field(default=None, max_length=100_000)
    slug: str | None = Field(default=None, max_length=1_000)
    url: HttpUrl | None = None
    canonical_url: HttpUrl | None = Field(default=None, alias="canonicalUrl")
    status: str | None = Field(default=None, max_length=100)
    brand: str | None = Field(default=None, max_length=500)
    product_type: str | None = Field(default=None, alias="productType", max_length=500)
    categories: list[str] = Field(default_factory=list, max_length=100)
    collections: list[str] = Field(default_factory=list, max_length=100)
    tags: list[str] = Field(default_factory=list, max_length=500)
    attributes: dict[str, JsonValue] = Field(default_factory=dict, max_length=500)
    seo: CatalogBridgeSeo = Field(default_factory=CatalogBridgeSeo)
    images: list[CatalogBridgeImage] = Field(default_factory=list, max_length=500)
    variants: list[CatalogBridgeVariant] = Field(default_factory=list, max_length=2_000)
    created_at: datetime | None = Field(default=None, alias="createdAt")
    updated_at: datetime | None = Field(default=None, alias="updatedAt")

    @field_validator("attributes")
    @classmethod
    def validate_attributes(cls, value: dict[str, JsonValue]) -> dict[str, JsonValue]:
        return _validate_catalog_value(value)  # type: ignore[return-value]

    @model_validator(mode="after")
    def validate_relationships(self) -> Self:
        variant_ids = [variant.id for variant in self.variants]
        if len(set(variant_ids)) != len(variant_ids):
            raise ValueError("Variant IDs must be unique inside a product")
        known_variants = set(variant_ids)
        if any(image.variant_id and image.variant_id not in known_variants for image in self.images):
            raise ValueError("Image variantId must reference a variant in the same product")
        return self


class CatalogBridgeSourceCreateRequest(CatalogBridgeModel):
    name: str = Field(min_length=2, max_length=200)


class CatalogBridgeSourceProvisionResponse(CatalogBridgeModel):
    source_id: uuid.UUID = Field(alias="sourceId")
    endpoint: HttpUrl
    token: str = Field(min_length=32)
    token_fingerprint: str = Field(alias="tokenFingerprint", min_length=12, max_length=12)
    protocol_version: str = Field(alias="protocolVersion")


class CatalogBridgeSnapshotManifest(CatalogBridgeModel):
    protocol_version: str = Field(alias="protocolVersion")
    snapshot_id: uuid.UUID = Field(alias="snapshotId")
    started_at: datetime = Field(alias="startedAt")
    declared_product_count: int = Field(alias="declaredProductCount", ge=0)
    declared_variant_count: int = Field(alias="declaredVariantCount", ge=0)
    source_label: str | None = Field(default=None, alias="sourceLabel", max_length=200)
    metadata: dict[str, str] = Field(default_factory=dict, max_length=30)

    @field_validator("protocol_version")
    @classmethod
    def validate_protocol(cls, value: str) -> str:
        if value != CATALOG_BRIDGE_PROTOCOL_VERSION:
            raise ValueError("Unsupported catalog bridge protocol version")
        return value


class CatalogBridgeBatch(CatalogBridgeModel):
    protocol_version: str = Field(alias="protocolVersion")
    snapshot_id: uuid.UUID = Field(alias="snapshotId")
    sequence: int = Field(ge=0)
    records: list[CatalogBridgeProduct] = Field(
        min_length=1,
        max_length=CATALOG_BRIDGE_MAX_BATCH_RECORDS,
    )

    @field_validator("protocol_version")
    @classmethod
    def validate_protocol(cls, value: str) -> str:
        if value != CATALOG_BRIDGE_PROTOCOL_VERSION:
            raise ValueError("Unsupported catalog bridge protocol version")
        return value


class CatalogBridgeCompleteRequest(CatalogBridgeModel):
    protocol_version: str = Field(alias="protocolVersion")
    snapshot_id: uuid.UUID = Field(alias="snapshotId")
    batch_count: int = Field(alias="batchCount", gt=0)
    product_count: int = Field(alias="productCount", ge=0)
    variant_count: int = Field(alias="variantCount", ge=0)

    @field_validator("protocol_version")
    @classmethod
    def validate_protocol(cls, value: str) -> str:
        if value != CATALOG_BRIDGE_PROTOCOL_VERSION:
            raise ValueError("Unsupported catalog bridge protocol version")
        return value


class CatalogBridgeSnapshotStatus(CatalogBridgeModel):
    source_id: uuid.UUID = Field(alias="sourceId")
    snapshot_id: uuid.UUID = Field(alias="snapshotId")
    status: str
    accepted_batches: int = Field(alias="acceptedBatches", ge=0)
    accepted_products: int = Field(alias="acceptedProducts", ge=0)
    accepted_variants: int = Field(alias="acceptedVariants", ge=0)
    ingestion_job_id: uuid.UUID | None = Field(default=None, alias="ingestionJobId")
